### PR TITLE
[FLINK-35016] catalog changes for model resource

### PR DESCRIPTION
**Pull request by lihaosky**
[## What is the purpose of the change

Catalog changes to support model resource

## Brief change log

* Add `CatalogModel` related resource
* Add model CRUD operations in `CatalogManager` and `Catalog`

## Verifying this change

Unit test added

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)
](https://media.giphy.com/media/KJ1f5iTl4Oo7u/giphy-downsized-large.gif)

### CODE REVIEW

File: flink-python/pyflink/table/catalog.py [hunk 1]
```diff
@@ -15,9 +15,8 @@
 #  See the License for the specific language governing permissions and
 # limitations under the License.
 ################################################################################
-from typing import Dict, List, Optional
-
 from py4j.java_gateway import java_import
+from typing import Dict, List, Optional
```
Comment (jnh5y) [inline flink-python/pyflink/table/catalog.py:19]: Did this line need to move down?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java [hunk 2]
```diff
@@ -888,7 +899,7 @@ public Set<String> listSchemas() {
      * <p><b>NOTE:</b>It is primarily used for interacting with Calcite's schema.
      *
      * @param catalogName filter for the catalog part of the schema
-     * @return list of schemas with the given prefix
+     * @return set of schemas with the given prefix
```
Comment (jnh5y) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java:902]: This change seems unrelated to models.  I wonder if we could pull them out into a purely cosmetic PR?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java [hunk 3]
```diff
@@ -1341,6 +1352,293 @@ private void dropTableInternal(
...
+                if (model != null) {
+                    final ResolvedCatalogModel resolvedModel = resolveCatalogModel(model);
+                    return Optional.of(
+                            ContextResolvedModel.permanent(
+                                    objectIdentifier, currentCatalog, resolvedModel));
+                }
+            } catch (ModelNotExistException e) {
+                // Ignore.
+            } catch (UnsupportedOperationException e) {
+                // Ignore for catalogs that don't support models.
```
Comment (jnh5y) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java:1383]: I wonder if it'd make sense for the catalog to advertise whether or not it supports models?  

Is that idea covered in the API defined by the FLIP?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java [hunk 4]
```diff
@@ -1341,6 +1352,293 @@ private void dropTableInternal(
...
+                objectIdentifier,
+                (k, v) -> {
+                    if (v != null) {
+                        if (!ignoreIfExists) {
+                            throw new ValidationException(
+                                    String.format(
+                                            "Temporary model '%s' already exists",
+                                            objectIdentifier));
+                        }
+                        return v;
```
Comment (jnh5y) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java:1493]: Does there need to be any check that the existing model and requested model share configuration/details other than just the name?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java [hunk 5]
```diff
@@ -476,11 +578,12 @@ public CatalogFunction getFunction(ObjectPath path) throws FunctionNotExistExcep
...
-        ObjectPath functionPath = normalize(path);
-
-        return databaseExists(functionPath.getDatabaseName())
-                && functions.containsKey(functionPath);
+        try {
+            getFunction(path);
+            return true;
+        } catch (Exception e) {
+            return false;
+        }
```
Comment (jnh5y) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java:586]: This also appears to be separate refactoring.

**Comment by jeff303** [comment]
This PR replaces #25036.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java [hunk 6]
```diff
@@ -1341,6 +1352,293 @@ private void dropTableInternal(
...
+    }
+
+    public ResolvedCatalogModel resolveCatalogModel(CatalogModel model) {
+        Preconditions.checkNotNull(schemaResolver, "Schema resolver is not initialized.");
+        if (model instanceof ResolvedCatalogModel) {
+            return (ResolvedCatalogModel) model;
+        }
+        // If input schema / output schema does not exist, get the schema from select as statement.
+        final ResolvedSchema resolvedInputSchema;
+        if (model.getInputSchema() == null) {
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java:1627]: Could you provide more information about this case? It looks like this is at the wrong layer. For tables, a catalog object has always a correct output type. Take a look at CREATE TABLE AS which has similar behavior.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/ContextResolvedModel.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    private ContextResolvedModel(
+            ObjectIdentifier objectIdentifier,
+            @Nullable Catalog catalog,
+            ResolvedCatalogModel resolvedModel) {
+        this.objectIdentifier = Preconditions.checkNotNull(objectIdentifier);
+        this.catalog = catalog;
+        this.resolvedModel = Preconditions.checkNotNull(resolvedModel);
+    }
+
+    /** @return true if the table is temporary. An anonymous table is always temporary. */
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/ContextResolvedModel.java:75]:
```suggestion
    /** @return true if the model is temporary. */
```

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/CatalogManagerTest.java [hunk 8]
```diff
@@ -23,16 +23,21 @@
...
 import org.apache.flink.table.catalog.listener.CreateDatabaseEvent;
+import org.apache.flink.table.catalog.listener.CreateModelEvent;
 import org.apache.flink.table.catalog.listener.CreateTableEvent;
 import org.apache.flink.table.catalog.listener.DropDatabaseEvent;
+import org.apache.flink.table.catalog.listener.DropModelEvent;
 import org.apache.flink.table.catalog.listener.DropTableEvent;
 import org.apache.flink.table.utils.CatalogManagerMocks;
 import org.apache.flink.table.utils.ExpressionResolverMocks;
 
+import org.apache.flink.shaded.guava32.com.google.common.collect.ImmutableMap;
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/CatalogManagerTest.java:39]: the use of Guava should be avoided, use a regular HashMap

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/CatalogManagerTest.java [hunk 9]
```diff
@@ -256,6 +262,114 @@ public void testDropCurrentDatabase() throws Exception {
...
 
+    @Test
+    public void testModelModificationListener() throws Exception {
+        CompletableFuture<CreateModelEvent> createFuture = new CompletableFuture<>();
+        CompletableFuture<CreateModelEvent> createTemporaryFuture = new CompletableFuture<>();
+        CompletableFuture<AlterModelEvent> alterFuture = new CompletableFuture<>();
+        CompletableFuture<DropModelEvent> dropFuture = new CompletableFuture<>();
+        CompletableFuture<DropModelEvent> dropTemporaryFuture = new CompletableFuture<>();
+        CatalogManager catalogManager =
+                CatalogManager.newBuilder()
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/CatalogManagerTest.java:273]: use `org.apache.flink.table.utils.CatalogManagerMocks#preparedCatalogManager`

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/CatalogManagerTest.java [hunk 10]
```diff
@@ -460,10 +617,16 @@ void testCatalogStore() throws Exception {
                         Collections.emptyMap()),
                 ObjectIdentifier.of("exist_cat", "cat_db", "test_table"),
                 false);
+        catalogManager.createModel(
+                CatalogModel.of(Schema.derived(), Schema.derived(), Collections.emptyMap(), null),
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/CatalogManagerTest.java:621]: I'm fine using Schema.derived here but then you should not perform null checks in CatalogManager. Because Schema.derived is not null.

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/GenericInMemoryCatalogTest.java [hunk 11]
```diff
@@ -52,6 +60,197 @@ static void init() {
...
+        if (catalog.modelExists(modelPath1)) {
+            ((GenericInMemoryCatalog) catalog).dropModel(modelPath1, true);
+        }
+        if (catalog.modelExists(modelPath2)) {
+            ((GenericInMemoryCatalog) catalog).dropModel(modelPath2, true);
+        }
+        super.cleanup();
+    }
+
+    // These are put here instead of CatalogTest class since model operations are not implemented
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/GenericInMemoryCatalogTest.java:75]: rather introduce a `supportsModels(): boolean` method in CatalogTest and move those tests to the testbase. Set this flag to false and ignore them for Hive catalog.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogModel.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+public interface CatalogModel {
+    /** Returns a map of string-based model options. */
+    Map<String, String> getOptions();
+
+    /**
+     * Get the unresolved input schema of the model.
+     *
+     * @return unresolved input schema of the model.
+     */
+    Schema getInputSchema();
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogModel.java:39]: So can this be null or not?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/DefaultCatalogModel.java [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    private final Schema outputSchema;
+    private final Map<String, String> modelOptions;
+    private final @Nullable String comment;
+
+    protected DefaultCatalogModel(
+            Schema inputSchema,
+            Schema outputSchema,
+            Map<String, String> modelOptions,
+            @Nullable String comment) {
+        this.inputSchema = inputSchema;
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/DefaultCatalogModel.java:42]: add null checks for the args to avoid nulls travelling through the stack

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogModel.java (hunk 12 above)
Comment (lihaosky) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogModel.java:39]: Cannot be null. Added null check

**Review by twalthr — APPROVED** [review]
Thanks for the update @lihaosky. LGTM
